[화살] 화살 연관 로직 수정(#68)

Minjae-An opened the pull request:
## 작업 대상
<!-- 작업한 대상을 자세히 설명해주세요. -->
- 출석 체크
- 화살 보내기
- 화살 충전

## 📄 작업 내용
<!-- 작업한 내용을 간단히 요약해주세요. -->
### 화살 내역 타입 추가
- 출석 체크, 화살 충전은 모두 `sender` 칼럼이 `null`로 저장됨
- 이 둘을 구분하기 위해 화살 타입(`ArrowTransactionType`) 열거형 정의, 내역에 포함

### 출석 체크
- 휴면 상태 사용자가 출석 체크할 수 없도록 검증
- 이미 출석 체크한 상황을 예외로 다루지 않음
- 출석 체크와 사용자 최종 접속 일시 갱신을 한 트랜잭션으로 다룰 수 있는 구조로 구성
- 출석 체크 비즈니스 로직이 출석 체크 진행 여부를 반환하여 controller에서 다른 형태의 API 응답 제공토록 로직 구성

### 화살 보내기
- 휴면 상태 사용자가 화살을 보내거나 받을 수 없도록 검증
- 같은 성별 사용자에게 화살을 보낼 수 없도록 검증

### 화살 충전
- 화살 타입 추가에 따라 일일 3회 충전 제한 검증 로직에서 타입을 고려하도록 수정

## 🙋🏻 주의 사항
<!-- 리뷰어를 위해 복잡하거나 중요한 코드를 명시해주세요. -->


## 📎 관련 이슈
- `Gender`를 DB엔 `boolean` 형태로 저장하지만 코드에서는 다른 형태로 활용하다 보니 로직이 부자연스러워짐
- 열거형 타입 그대로를 코드에서 활용할 수 있도록 추후 리팩터링 예정

## 레퍼런스
<!-- 작업 시 참고했던 문건의 URL을 남겨주세요 -->

Minjae-An committed:
feat: lombok 활용하도록 변경(#68)

Minjae-An committed:
feat: gender boolean 값 string 변환 메서드 추가(#68)

Minjae-An committed:
feat: 사용자 엔티티 수정(#68)

- 같은 사용자(자기 자신) 검증 로직, UserException 이용토록 수정
- 성별 boolean, string  두 타입으로 제공하는 메서드 추가
- 같은 성별 사용자인 지 검증하는 메서드 추가

Minjae-An committed:
feat: 사용자 연관 예외 추가(#68)

- 같은 사용자(자기 자신)일 경우
- 같은 성별 사용자인 경우
- 휴면 상태 사용자인 경우

Minjae-An committed:
feat: 사용자 엔티티 boolean 타입 성별 getter 사용 위치 수정(#68)

- getter 메서드명 변경에 따른 사용위치 수정

Minjae-An committed:
feat: 휴면 상태 사용자에게 화살 보냄 예외 정의(#68)

Minjae-An committed:
feat: 불필요한 화살 보내기 요청 DTO 삭제(#68)

Minjae-An committed:
feat: 화살 보내기 비즈니스 로직 수정(#68)

- 보내는 사용자 휴면 상태 검증 추가
- 받는 사용자 휴면 상태 검증 추가
- 보내는 사용자와 받는 사용자 성별 동일 여부 검증 추가
- 화살 개수 1개만 보낼 수 있도록 수정

Minjae-An committed:
feat: 화살 보내기 API 수정(#68)

한 번에 하나의 화살만 보낼 수 있기 때문에 별도의 요청 DTO(request body)
없이 작업을 수행하도록 API 수정

Minjae-An committed:
feat: 보내는 사용자, 받는 사용자 필드명 수정(#68)

Minjae-An committed:
feat: 화살 내역 엔티티 필드명 변경에 따른 사용 위치 수정(#68)

Minjae-An committed:
refactor: 패키지 경로 이동(#68)

Minjae-An committed:
feat: 최종 접속 일시 갱신 메서드 파라미터명 변경(#68)

Minjae-An committed:
feat: 최종 접속 일시 갱신 로직 추가(#68)

출석 체크 요청시 무조건적인 최종 접속 일시 갱신을 위해 사용

Minjae-An committed:
feat: 인증되지 않은 사용자 예외 추가(#68)

Minjae-An committed:
feat: 최종 접속 일시 갱신 인터셉터 추가(#68)

- 출석 체크 성공/실패 여부와 상관 없이 사용자 최종 접속 일시는 무조건 갱신
- 이 인터셉터를 통해 출석 체크 요청 이후 무조건적으로 최종 접속 일시를 업데이트 처리

Minjae-An committed:
feat: 최종 접속 일시 갱신 인터셉터 등록 로직 추가(#68)

- 출석 체크 요청시에만 최종 접속 일시 갱신을 수행하도록 인터셉터 등록

Minjae-An committed:
feat: 출석 체크 비즈니스 로직 수정(#68)

- 최종 접속 일시 갱신을 별도의 인터셉터가 수행하도록 분리
- 휴면 상태 사용자는 출석 체크 불가하도록 검증
- 처음 가입한 사용자는 최종 접속 일시가 null이므로 접속 일시 검증을 수행하지 않음

Minjae-An committed:
feat: 사용자 화살 개수 응답 DTO 사용 위치 수정(#68)

Minjae-An committed:
feat: 메서드 파라미터명 수정(#68)

Minjae-An committed:
feat: 출석 체크 로직, 출석 체크 날짜 파라미터 추가(#68)

테스트 가능한 형태로 외부에서 출석 체크 날짜를 주입받도록 수정

Minjae-An committed:
feat: 출석 체크 API 로직, 출석 체크 날짜 생성 코드 추가(#68)

Minjae-An committed:
feat: 화살 충전 로직, 충전일 파라미터 추가(#68)

- 기존 로직은 로직 실행일 기준으로 충전 최대 횟수 검증 수행
- 테스트하기 어려운 구조, 외부에서 충전일을 전달받도록 변경

Minjae-An committed:
feat: 화살 충전 API, 충전일 생성 코드 추가(#68)

화살 충전 버즈니스 로직이 충전일을 외부에서 제공받는 형태로
변경됨에 따라 충전일 생성하는 코드 추가

Minjae-An committed:
Merge branch 'dev' into feat/#68-daily-check-and-send-arrow

Minjae-An committed:
feat: 최종 접속 일시 업데이트 로직, 간단한 형태로 수정(#68)

Minjae-An committed:
feat: 화살 내역 타입 추가(#68)

화살 내역을 경우에 따라 구분하기 위해 추가한 열거형, 다음 종류들이 존재
- 출석 체크
- 광고 시청을 통한 화살 획득
- 사용자간 송수신

Minjae-An committed:
feat: 화살 내역 타입 필드 추가(#68)

Minjae-An committed:
feat: 광고 시청 가능 여부 확인 로직, 화살 내역 타입 비교 조건 추가(#68)

Minjae-An committed:
feat: 화살 연관 로직들 수정(#68)

- 화살 내역 생성시 화살 내역 타입을 함께 포함하여 내역마다 구분이 가능하도록 수정
- 불필요한 상수, 지역 변수로 구성
- 의미가 더 잘 드러나도록 변수명 적절히 수정

Minjae-An committed:
Merge branch 'dev' into feat/#68-daily-check-and-send-arrow

acceptor-gyu commented on be/src/main/java/yeonba/be/arrow/service/ArrowService.java:
```diff
@@ -1,52 +1,56 @@
+            dailyCheckArrows);
         arrowCommand.save(arrowTransaction);
 
-        dailyCheckUser.updateLastAccessedAt(dailyCheckedAt);
```
최종 접속 일시 수정을 interceptor에서 하신 이유가 무엇인가요?
- 출석체크 화살 추가와 최종 접속 일시 수정은 하나의 트랜잭션에서 동작해야 할 것같은데 이대로 문제가 없을까요?
- 수정하신 로직으로는 화살 추가는 됐지만, 최종 접속 일시는 수정이 되지 않을 가능성이 있을 것같은데 제가 잘못 이해한 건가요?

Minjae-An replied:
- 출석체크가 시도될 경우, 출석체크 정상 처리여부와 관계없이 최종 접속 일시는 갱신되어야 합니다.
- 따라서 두 작업을 한 트랜잭션으로 다루지 않았으며, 출석체크 로직에서 예외가 발생하더라도 최종 접속 일시는 
  갱신되도록 인터셉터의 `afterCompletion`을 활용했습니다.
- 화살 추가는 됐지만, 최종 접속 일시가 수정되지 않을 상황이 최종 접속 일시 로직의 예외 상황을 말씀하시는 
  것일까요? 잘 이해하지 못해 질문드립니다.

acceptor-gyu replied:
네네 사용자에게 화살은 추가됐지만 최종 접속 시간은 업데이트 되지 않은 경우는 정상적인 상황이 아닌 것같아서요.

Minjae-An replied:
두 작업을 한 트랜잭션으로 다룰 수 있게 변경하였습니다. 이미 출석 체크한 상황을 예외로 다루지 않는 형식으로 로직을 구성하였는데 적절한 지 한 번 피드백 주시길 부탁드립니다.

Minjae-An committed:
feat: 사용하지 않는 최종 접속 일시 업데이트 인터셉터 삭제(#68)

Minjae-An committed:
feat: 사용하지 않는 예외 삭제(#68)

Minjae-An committed:
feat: 코드 정렬(#68)

Minjae-An committed:
feat: 출석 체크 가능 여부 확인 로직 수정(#68)

- 기존 로직에서는 이미 출석 체크한 상황을 예외로 다룸
- 사용자 화살 증가와 최종 접속 일시 갱신을 한 트랜잭션으로 다루기 어려운 구조
- 이미 출석 체크한 상황을 예외로 다루지 않고, 출석 체크일과 최종 접속 일시 비교 결과 반환
- 출석 체크 로직에서 비교 결과에 따라 진행 토록 로직 변경

Minjae-An committed:
feat: 사용하지 않는 최종 접속 일시 업데이트 로직 삭제(#68)

Minjae-An committed:
feat: 출석 체크 로직 수정(#68)

- 출석 체크 내역 저장, 사용자 화살 증가와 최종 접속 일시 갱신을 한 트랜잭션으로 다루도록 변경
- 이를 위해 이미 출석 체크한 상황을 예외로 다루지 않음
- 출석 체크 진행 여부를 반환하여 출석 체크 진행 여부를 판단하도록 구성

Minjae-An committed:
feat: 출석 체크 API 로직 수정(#68)

- 서비스 로직을 출석 체크 진행 여부를 반환하도록 변경
- 결과 값에 따라 적절한 응답을 클라이언트에 전달

Minjae-An committed:
feat: 사용되지 않는 예외 삭제(#68)

acceptor-gyu reviewed (approved):
시연이 급해 approve 하겠습니다. 이후 버그 발생 시 수정부탁드립니다.
